pkg/queue/health/probe.go

```diff
@@ -202,3 +219,69 @@ func isHTTPProbeReady(res *http.Response) bool {
 	// response status code between 200-399 indicates success
 	return res.StatusCode >= 200 && res.StatusCode < 400
 }
+
+// GRPCProbe checks that gRPC connection can be established to the address.
+func GRPCProbe(config GRPCProbeConfigOptions) error {
+	dialer := &net.Dialer{
+		Control: func(network, address string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
+				unix.SetsockoptLinger(int(fd), syscall.SOL_SOCKET, syscall.SO_LINGER, &unix.Linger{Onoff: 1, Linger: 1})
+			})
+		},
+	}
+
+	v := version.Get()
+
+	opts := []grpc.DialOption{
+		grpc.WithUserAgent(fmt.Sprintf("kube-probe/%s.%s", v.Major, v.Minor)),
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()), // credentials are currently not supported
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			return dialer.DialContext(ctx, "tcp", addr)
+		}),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), config.Timeout)
+
+	defer cancel()
+
+	addr := net.JoinHostPort("127.0.0.1", fmt.Sprintf("%d", config.Port))
+	conn, err := grpc.DialContext(ctx, addr, opts...)
+
+	if err != nil {
+		if err == context.DeadlineExceeded {
+			return fmt.Errorf("timeout: failed to connect service %q within %v", addr, config.Timeout)
+		} else {
+			return fmt.Errorf("error: failed to connect service at %q", addr)
+		}
+	}
+
+	defer func() {
+		_ = conn.Close()
+	}()
+
+	client := grpchealth.NewHealthClient(conn)
+
+	resp, err := client.Check(metadata.NewOutgoingContext(ctx, make(metadata.MD)), &grpchealth.HealthCheckRequest{
+		Service: ptr.ToString(config.Service),
```

Comment:
Don't need to pull in a new pkg - we have https://github.com/knative/pkg/blob/main/ptr/ptr.go#L53